contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+			)
+			.with_gas_limit(action.gas_limit);
+
+			if action.amount > 0 {
+				contract_call = contract_call
+					.with_token_transfer(action.token_id, action.amount)
+					.with_nft_nonce(action.token_nonce);
+			}
+
+			for arg in action.arguments {
+				contract_call.push_argument_raw_bytes(arg.as_slice());
+			}
+
+			contract_call.transfer_execute();
+		}
+
+		self.clear_proposal(proposal_id);
+
+		Ok(())
+	}
+
+	#[endpoint]
+	fn cancel(&self, proposal_id: usize) -> SCResult<()> {
+		match self.get_proposal_status(proposal_id) {
+			GovernanceProposalStatus::None => {
+				return sc_error!("Proposal does not exist");
+			},
+			GovernanceProposalStatus::Defeated => {},
+			_ => {
+				let proposal = self.proposals().get(proposal_id);
+				let caller = self.blockchain().get_caller();
+
+				require!(
+					caller == proposal.proposer,
+					"Only original proposer may cancel a non-defeated proposal"
+				);
+			},
+		}
+
+		self.clear_proposal(proposal_id);
+
+		Ok(())
+	}
+
+	// views
+
+	#[view(getProposalStatus)]
+	fn get_proposal_status(&self, proposal_id: usize) -> GovernanceProposalStatus {
+		if !self.proposal_exists(proposal_id) {
+			return GovernanceProposalStatus::None;
+		}
+
+		let queue_block = self.proposal_queue_block(proposal_id).get();
+		if queue_block > 0 {
+			return GovernanceProposalStatus::Queued;
+		}
+
+		let current_block = self.blockchain().get_block_nonce();
+		let proposal_block = self.proposal_start_block(proposal_id).get();
+		let voting_delay = self.voting_delay_in_blocks().get();
+		let voting_period = self.voting_period_in_blocks().get();
+
+		let voting_start = proposal_block + voting_delay;
+		let voting_end = voting_start + voting_period;
+
+		if current_block < voting_start {
+			return GovernanceProposalStatus::Pending;
+		}
+		if current_block >= voting_start && current_block < voting_end {
+			return GovernanceProposalStatus::Active;
+		}
+
+		let total_votes = self.total_votes(proposal_id).get();
+		let total_downvotes = self.total_downvotes(proposal_id).get();
+		let quorum = self.quorum().get();
+
+		if total_votes > total_downvotes && total_votes - total_downvotes >= quorum {
+			return GovernanceProposalStatus::Succeeded;
+		} else {
+			return GovernanceProposalStatus::Defeated;
+		}
+	}
+
+	#[view(getProposer)]
+	fn get_proposer(&self, proposal_id: usize) -> OptionalArg<Address> {
+		if !self.proposal_exists(proposal_id) {
+			OptionalArg::None
+		} else {
+			OptionalArg::Some(self.proposals().get(proposal_id).proposer)
+		}
+	}
+
+	#[view(getProposalDescription)]
+	fn get_proposal_description(&self, proposal_id: usize) -> OptionalArg<BoxedBytes> {
+		if !self.proposal_exists(proposal_id) {
+			OptionalArg::None
+		} else {
+			OptionalArg::Some(self.proposals().get(proposal_id).description)
+		}
+	}
+
+	#[view]
+	fn get_proposal_actions(
+		&self,
+		proposal_id: usize,
+	) -> MultiResultVec<GovernanceActionAsMultiArg<Self::BigUint>> {
+		if !self.proposal_exists(proposal_id) {
```

Comment:
Replace with require

Reply:
Cannot use `require!`, as this view function does not return `SCResult<T>` (which is intended, I prefer returning "empty" instead of error for view functions)